gregBorchers/HW5Problem3.js

```diff
@@ -0,0 +1,164 @@
+
+        var user = {
+            // "en-closured" methods
+            username: function() {
+                return newUserName;
+            },
+            pwd: function() {
+                return newPassword;
+            },
+            // references to methods "one scope outside" of the closure
+            getName: makeUser.getName,
+            getPwd: makeUser.getPwd,
+            validate: makeUser.validate,
+            record: makeUser.record
+        }; // end of closure
+
+        return user;
+    }
+
+
+
+    // shared methods for the user
+    makeUser.getName = function() {
+        return (this.username());
+    };
+
+    makeUser.validate = function(guess) { // TODO finish this
+        return (guess === this.pwd() ? "CORRECT -- good guess" : "WRONG -- you are stoopid hackerr");
+    };
+
+    makeUser.getPwd = function() { // TODO remove this before production :-) bad idea, but helped with testing
+        return ('password is ' + this.pwd());
+    };
+
+    makeUser.record = function(msg) {
+        userMsgLog.push('-->' + this.getName() + ' logged a message: ' + msg + '\n');
+    };
+
+    // not in the closure, this is a factory method
+    makeUser.getLog = function(logUserName) {
+        // check for empty args
+        var outStr = "Stoopid Hackerr, no logs for you!"; // the default answer
+
+        // check for empty args to trigger the full log case
+        if (arguments.length === 0) {
+            outStr = "START of log file for everything \n";
+            outStr += userMsgLog.join('\n');
+        } else { // get log records for one user
+            outStr = "START of log fileFor username = " + logUserName + '\n';
+
+            for (var line in userMsgLog) {
+                if (userMsgLog[line].indexOf(logUserName) >= 0) { // FOUND the user name in this log record
+                    outStr += userMsgLog[line] + '\n';
+                }
+            }
+        }
+        outStr += "END of log file\n";
+        return outStr;
+    };
+
+    return makeUser;
+
+}
+
+
+
+
+
+// testing to see if the basic structure works for getters --------------------------
+var makeUserFn = makeUserFactory(); // semantics of problem 3 start to fall apart here - using userFactory.makeUser() would be nicer
+var user1 = makeUserFn("bob", "sekret");
+
+// what I wanted to do:  
+//Still not sure why this did not work  
+//TODO later go back and figure out why these semantics don't work - had to get over it and move on
+// var userFactory = makeUserFactory();
+// var user1 = userFactory.makeUser("bob", "sekret");
+
+console.log("___ START ------- basic getter testing");
+console.log("this is all that I can see from outside of makeUser: " + Object.keys(user1));
+
+console.log("these dot access attempts should FAIL to show any data values.");
+console.log("there is no visibility into the ''en-closured'' parts via ''dot'' notation");
+console.log(user1.username);
+console.log(user1.pwd);
```

Comment:
Remember that these are functions, so you need to call them as such, like `console.log(user1.username())`. Doing so actually does return the username and password, which I'll comment on above in the relevant code.